datafusion/physical-expr/src/intervals/interval_aritmetic.rs

```diff
@@ -0,0 +1,533 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Interval arithmetic library
+
+use std::borrow::Borrow;
+use std::fmt;
+use std::fmt::{Display, Formatter};
+
+use arrow::compute::{cast_with_options, CastOptions};
+use arrow::datatypes::DataType;
+use datafusion_common::{DataFusionError, Result, ScalarValue};
+use datafusion_expr::Operator;
+
+use crate::aggregate::min_max::{max, min};
+
+/// This type represents an interval, which is used to calculate reliable
+/// bounds for expressions. Currently, we only support addition and
+/// subtraction, but more capabilities will be added in the future.
+/// Upper/lower bounds having NULL values indicate an unbounded side. For
+/// example; [10, 20], [10, ∞], [-∞, 100] and [-∞, ∞] are all valid intervals.
+#[derive(Debug, PartialEq, Clone, Eq, Hash)]
+pub struct Interval {
+    pub lower: ScalarValue,
+    pub upper: ScalarValue,
```

Comment:
Longer term to support partition pruning, supporting Interval arithmetic on `Array`s rather than `ScalarValue`s might be good / lower the cost to evaluate them.